types/wlr_surface.c

```diff
@@ -400,24 +401,33 @@ static void surface_update_input_region(struct wlr_surface *surface) {
 		0, 0, surface->current.width, surface->current.height);
 }
 
-static void surface_commit_pending(struct wlr_surface *surface) {
-	if (!surface_state_finalize(surface, &surface->pending)) {
+static void surface_state_init(struct wlr_surface_state *state);
+
+static void surface_cache_pending(struct wlr_surface *surface) {
+	struct wlr_surface_state *cached = calloc(1, sizeof(*cached));
+	if (!cached) {
+		wl_resource_post_no_memory(surface->resource);
 		return;
 	}
 
-	if (surface->role && surface->role->precommit) {
-		surface->role->precommit(surface);
-	}
+	surface_state_init(cached);
+	surface_state_move(cached, &surface->pending);
+	cached->cached_locks = surface->pending.cached_locks;
+	surface->pending.cached_locks = 0;
```

Comment:
I was originally worried about incorrectly doing it elsewhere because `surface_state_move` was called from many places. Looking at it again, this sounds fine.

While at it, I've also added a new assertion in `surface_commit_state` to make sure there are no cached state locks.